### PR TITLE
ROX-12966: disable node scanning for non RHCOS nodes

### PR DESCRIPTION
**Pull request by daynewlee**

### CODE REVIEW

**Comment by daynewlee** [comment]
Tested on GKE cluster:


```
time="2023-01-16T16:23:21Z" level=warning msg="Not able to start node scanning for this namespace" detected namespace="ubuntu:20.04" layer=gke-yl-01-16-relieved-en-default-pool-49b42e9d-2hbn
collection/nodeinventorizer: 2023/01/16 16:23:21.145016 nodeinventory.go:27: Info: Finished node inventory
collection/nodeinventorizer: 2023/01/16 16:23:21.145055 nodeinventory.go:29: Error: Error scanning node /host inventory: Node scanning unavailable
main: 2023/01/16 16:23:21.145141 main.go:173: Error: error running scanNode: Node scanning unavailable
collection/nodeinventorizer: 2023/01/16 16:24:20.839038 nodeinventory.go:23: Info: Started node inventory
```

**Comment by daynewlee** [comment]
Tested on Openshift cluster:

```
main: 2023/01/16 18:35:43.554169 main.go:243: Info: Successfully connected to Sensor at sensor.stackrox.svc:443
collection/nodeinventorizer: 2023/01/16 18:35:49.130227 nodeinventory.go:27: Info: Finished node inventory
main: 2023/01/16 18:35:49.146348 main.go:243: Info: Successfully connected to Sensor at sensor.stackrox.svc:443
collection/nodeinventorizer: 2023/01/16 18:36:43.461517 nodeinventory.go:23: Info: Started node inventory
time="2023-01-16T18:36:43Z" level=info msg="add files from directory" directory=etc/ root=/host
time="2023-01-16T18:36:43Z" level=info msg="add files from directory" directory=usr/share/rpm root=/host
time="2023-01-16T18:36:43Z" level=info msg="add files from directory" directory=var/lib/rpm root=/host
collection/nodeinventorizer: 2023/01/16 18:36:47.538836 nodeinventory.go:27: Info: Finished node inventory

```

File: pkg/analyzer/detection/detection.go
Comment (RTann) [inline pkg/analyzer/detection/detection.go:25]: I think it may be better to have something like `DetectComponentOpts` which has attributes `uncertifiedRHEL` and maybe a slice of supported namespaces. This way, we don't have to keep adding parameters to the function. It also makes it more explicit what the parameters mean

File: pkg/analyzer/detection/detection.go
Comment (vikin91) [inline pkg/analyzer/detection/detection.go:38]: Will this error lead to setting a Note so that the user is informed that we do not support node scanning, or would that be a part of a follow-up?

File: pkg/analyzer/nodes/node.go
Comment (vikin91) [inline pkg/analyzer/nodes/node.go:70]: `IsRHCOSRequired: false` is hardcoded here, but we want to have ability to set it to `true` when we call `Analyze` from Compliance, right?

File: pkg/analyzer/detection/detection.go
Comment (daynewlee) [reply pkg/analyzer/detection/detection.go:38]: > Will this error lead to setting a Note so that the user is informed that we do not support node scanning, or would that be a part of a follow-up?

My current plan is to log and return an error in the compliance level. If the node scanning stops, I am not sure if we will still need to generate the node inventory.

File: pkg/analyzer/detection/detection.go
Comment (RTann) [inline pkg/analyzer/detection/detection.go:19]: at the moment yes, but `opts` implies these are just options we can set to configure how we want the detection to work

File: pkg/analyzer/detection/detection.go
Comment (RTann) [inline pkg/analyzer/detection/detection.go:24]: what happens if it's false?

File: pkg/analyzer/nodes/node.go
Comment (RTann) [inline pkg/analyzer/nodes/node.go:59]: not sure if this is overkill, but I think technically we should probably have a separate set of options for `Analysis` which currently, coincidentally matches the options in `detection.DetectComponentOpts`

Comment (vikin91) [reply pkg/analyzer/nodes/node.go:59]: This is okay, but I personally would prefer having a separate options type (i.e. not reuse `detection.DetectComponentOpts`) as Ross suggested.

File: pkg/analyzer/detection/detection.go
Comment (vikin91) [inline pkg/analyzer/detection/detection.go:25]: I would be a bit more explicit here. My suggestion:

```suggestion
	// StopEarlyForNonRHCOS: if true, the analysis will run only for namespace starting with `rhcos:`. If false, the analysis will stop after detecting the namespace.
	StopEarlyForNonRHCOS bool
```

**Review by vikin91 — APPROVED** [review]
This looks good, my comments are optional to address.
I would suggest to maybe get a green review from Scanner team member before merging.

I think, we should also discuss with PM the topic of setting a Note when non-RHCOS system is detected.

File: pkg/analyzer/detection/detection.go
Comment (Maddosaurus) [reply pkg/analyzer/detection/detection.go:25]: Nit: I found the naming of `IsRHCOSRequired` a bit confusing as well. 
If I read this as consumer of the lib, I would ask myself "in which case would RHCOS be required?".

@vikin91's proposal seems fine to me. Alternatively I would pitch something like `RHCOSMode`  or `StrictMode`, although I'm not strongly in favor of any of these 🙂

File: pkg/analyzer/detection/detection.go
Comment (Maddosaurus) [inline pkg/analyzer/detection/detection.go:34]: Nit: Should we maybe update the comment above to include the info about the early exit on non-RHCOS namespaces if the scan opts are set?

**Review by Maddosaurus — APPROVED** [review]
Thanks for the change! Feel free to TAL at the small, optional things that came to mind while reading this change.

**Comment by vikin91** [comment]
@daynewlee one more question, it would be nice to edit the PR description to include there a sentence or two about how you tested this change. We want to make sure that it works after merging ;)

**Comment by vikin91** [comment]
Could we also fix style before merging? (see "style-checks" in CI)

**Pull request merged**
